### PR TITLE
Lazy clients, client DI, and expose client health check

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Added `TemporalClient.CreateLazy` and `TemporalConnection.CreateLazy` for those wanting to create client but not connect until later
* Added `ITemporalConnection.IsConnected` property and `ITemporalConnection.ConnectAsync` instance method to support lazy connections
* Added `IServiceCollection.AddTemporalClient` extension methods for creating DI-based lazy singleton client (one returns an option builder, the other accepts an action to configure)
* Added gRPC health check and supporting gRPC messages

### CODE REVIEW

**Comment by mrtristan** [comment]
looks good, thanks!

**Comment by cretz** [comment]
Note, I cannot merge this until I have released the existing in-limbo packages (there is a NuGet bug holding this up).

**Comment by alrz** [comment]
> Note, I cannot merge this until I have released the existing in-limbo packages (there is a NuGet bug holding this up).

I suppose you're referring to hosting/otel packages, any eta on releasing those to nugetorg? Is there an issue tracking nuget bug?

**Comment by cretz** [comment]
> I suppose you're referring to hosting/otel packages, any eta on releasing those to nugetorg? Is there an issue tracking nuget bug?

Hopefully soon, no ETA.

File: src/Temporalio/Client/TemporalConnection.cs [hunk 1]
```diff
@@ -12,13 +17,14 @@ namespace Temporalio.Client
     /// </summary>
     public sealed class TemporalConnection : ITemporalConnection
     {
-        private readonly Bridge.Client client;
+        // Not set if not lazy
+        private readonly SemaphoreSlim? semaphoreForLazyClient;
```
Comment (Sushisource) [inline src/Temporalio/Client/TemporalConnection.cs:21]: SemaphoreSlim sounds like a great rapper name

**Review by Sushisource — APPROVED** [review]

**Pull request merged**